server/src/apub/fetcher.rs

```diff
@@ -109,50 +58,114 @@ pub fn search_by_apub_id(query: &str, conn: &PgConnection) -> Result<SearchRespo
+) -> Result<User_, Error> {
+  match User_::read_from_actor_id(&conn, &apub_id) {
+    Ok(u) => {
+      // If its older than a day, re-fetch it
+      // TODO the less than needs to be tested
+      if u
+        .last_refreshed_at
+        .lt(&(naive_now() - chrono::Duration::days(1)))
+      {
+        debug!("Fetching and updating from remote user: {}", apub_id);
+        let person = fetch_remote_object::<PersonExt>(&Url::parse(apub_id)?)?;
+        let mut uf = UserForm::from_apub(&person, &conn)?;
+        uf.last_refreshed_at = Some(naive_now());
+        Ok(User_::update(&conn, u.id, &uf)?)
+      } else {
+        Ok(u)
+      }
+    }
+    Err(NotFound {}) => {
+      debug!("Fetching and creating remote user: {}", apub_id);
+      let person = fetch_remote_object::<PersonExt>(&Url::parse(apub_id)?)?;
+      let uf = UserForm::from_apub(&person, &conn)?;
+      Ok(User_::create(conn, &uf)?)
+    }
+    Err(e) => Err(Error::from(e)),
+  }
 }
 
-/// Fetch a user, insert/update it in the database and return the user.
-pub fn fetch_remote_user(apub_id: &Url, conn: &PgConnection) -> Result<User_, Error> {
-  let person = fetch_remote_object::<PersonExt>(apub_id)?;
-  let uf = UserForm::from_person(&person)?;
-  upsert_user(&uf, conn)
+/// Check if a remote community exists, create if not found, if its too old update it.Fetch a community, insert/update it in the database and return the community.
+pub fn get_or_fetch_and_upsert_remote_community(
+  apub_id: &str,
+  conn: &PgConnection,
+) -> Result<Community, Error> {
+  match Community::read_from_actor_id(&conn, &apub_id) {
+    Ok(c) => {
+      // If its older than a day, re-fetch it
+      // TODO the less than needs to be tested
+      if c
+        .last_refreshed_at
+        .lt(&(naive_now() - chrono::Duration::days(1)))
+      {
+        debug!("Fetching and updating from remote community: {}", apub_id);
+        let group = fetch_remote_object::<GroupExt>(&Url::parse(apub_id)?)?;
+        let mut cf = CommunityForm::from_apub(&group, conn)?;
+        cf.last_refreshed_at = Some(naive_now());
+        Ok(Community::update(&conn, c.id, &cf)?)
+      } else {
+        Ok(c)
+      }
+    }
+    Err(NotFound {}) => {
+      debug!("Fetching and creating remote community: {}", apub_id);
+      let group = fetch_remote_object::<GroupExt>(&Url::parse(apub_id)?)?;
+      let cf = CommunityForm::from_apub(&group, conn)?;
+      Ok(Community::create(conn, &cf)?)
+    }
+    Err(e) => Err(Error::from(e)),
+  }
 }
 
-/// Fetch a community, insert/update it in the database and return the community.
-pub fn fetch_remote_community(apub_id: &Url, conn: &PgConnection) -> Result<Community, Error> {
-  let group = fetch_remote_object::<GroupExt>(apub_id)?;
-  let cf = CommunityForm::from_group(&group, conn)?;
-  upsert_community(&cf, conn)
-}
+// TODO Maybe add post, comment searching / caching at a later time
+// fn upsert_post(post_form: &PostForm, conn: &PgConnection) -> Result<Post, Error> {
```

Comment:
I'm just commenting this because it *might* be useful at a later time to search for posts, comments, and just insert them if we don't have them. But I don't wanna add this till after we get proper inbox posting fully working, because we might confuse ourselves.

Reply:
Sorry but I dont see what this has to do with the inbox?

Reply:
If we have methods to upsert posts on searching an outbox (not really the activitypub way), but also methods where a post should be inserted when we receive an activity (correct activitypub), then there might be some ambiguity as to whether we did it the correct way. We should get the correct way working first before we decide if we ever want to grab history (and not just post history, but all history, comments, comment_likes, etc) from a community's outbox.